sdk/core/core-client/src/deserializationPolicy.ts

```diff
@@ -0,0 +1,280 @@
+  /**
+   * The content-types that indicate that an operation response should be deserialized as JSON.
+   * Defaults to [ "application/json", "text/json" ].
+   */
+  json?: string[];
+
+  /**
+   * The content-types that indicate that an operation response should be deserialized as XML.
+   * Defaults to [ "application/xml", "application/atom+xml" ].
+   */
+  xml?: string[];
+}
+
+/**
+ * This policy handles parsing out responses according to OperationSpecs on the request.
+ */
+export function deserializationPolicy(options: DeserializationPolicyOptions = {}): PipelinePolicy {
+  const jsonContentTypes = options.expectedContentTypes?.json ?? defaultJsonContentTypes;
+  const xmlContentTypes = options.expectedContentTypes?.xml ?? defaultXmlContentTypes;
+  const parseXML = options.parseXML;
+
+  return {
+    name: deserializationPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      const response = await next(request);
+      return deserializeResponseBody(jsonContentTypes, xmlContentTypes, response, parseXML);
+    }
+  };
+}
+
+function getOperationResponseMap(
+  parsedResponse: PipelineResponse
+): undefined | OperationResponseMap {
+  let result: OperationResponseMap | undefined;
+  const request: OperationRequest = parsedResponse.request;
+  const operationSpec = request.operationSpec;
+  if (operationSpec) {
+    if (!request.operationResponseGetter) {
+      result = operationSpec.responses[parsedResponse.status];
+    } else {
+      result = request.operationResponseGetter(operationSpec, parsedResponse);
+    }
+  }
+  return result;
+}
+
+function shouldDeserializeResponse(parsedResponse: PipelineResponse): boolean {
+  const request: OperationRequest = parsedResponse.request;
+  const shouldDeserialize = request.shouldDeserialize;
+  let result: boolean;
+  if (shouldDeserialize === undefined) {
+    result = true;
+  } else if (typeof shouldDeserialize === "boolean") {
+    result = shouldDeserialize;
+  } else {
+    result = shouldDeserialize(parsedResponse);
+  }
+  return result;
+}
+
+async function deserializeResponseBody(
+  jsonContentTypes: string[],
+  xmlContentTypes: string[],
+  response: PipelineResponse,
+  parseXML?: (str: string, opts?: { includeRoot?: boolean }) => Promise<any>
+): Promise<PipelineResponse> {
+  const parsedResponse = await parse(jsonContentTypes, xmlContentTypes, response, parseXML);
+  if (!shouldDeserializeResponse(parsedResponse)) {
+    return parsedResponse;
+  }
+
+  const operationSpec = parsedResponse.request.operationSpec;
+  if (!operationSpec || !operationSpec.responses) {
+    return parsedResponse;
+  }
+
+  const responseSpec = getOperationResponseMap(parsedResponse);
+  const expectedStatusCodes = Object.keys(operationSpec.responses);
+  const hasNoExpectedStatusCodes =
+    expectedStatusCodes.length === 0 ||
+    (expectedStatusCodes.length === 1 && expectedStatusCodes[0] === "default");
```

Comment:
Just a note, and probably doesn't make sense to add support until we switch to the new way of serializing/deserializing.

Autorest supports `x-ms-error-response` which allows users to mark responses as errors in Swagger and still be able to provide alternate deserialization rules to "default", currently we'd just see those as just another success response.

Reply:
yeah this logic is *terrible* and unchanged from the old stuff. I'm fine adding support here, but maybe you can file an issue to do it in a follow-up PR that is more scoped to the behavior change?